File: ios/Counter.m
//
//  Counter.m
//  JS_Proj_69
//
//  Created by Saunak on 18/02/23.
//



// RCT_EXTERN_MODULE macro - this is for exposing the Counter to JS class.

#import "React/RCTBridgeModule.h"
#import "React/RCTEventEmitter.h"

/*
 
 @interface RCT_EXTERN_MODULE( Counter, NSObject)
 @end
 */


/*
 For giving the class a new name we can use below macro
 
  First argument - Name that will be exposed to JS
  2nd argument - Swift Class
  3rd argument - super class
 */

@interface RCT_EXTERN_REMAP_MODULE(RNCounter, Counter, RCTEventEmitter)
RCT_EXTERN_METHOD(increment)
RCT_EXTERN_METHOD(getCount: (RCTResponseSenderBlock)callback)
RCT_EXTERN_METHOD(
                  decrement: (RCTPromiseResolveBlock) resolve
                  rejecter: (RCTPromiseRejectBlock) reject
                  )
@end

File: ios/Counter.swift
//
//  Counter.swift
//  JS_Proj_69
//
//  Created by Saunak on 18/02/23.
//

import Foundation

@objc(Counter)
class Counter: RCTEventEmitter {
  
  private var count = 0
  
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  
  override func constantsToExport () -> [AnyHashable: Any]!{
    return ["initialCount": 0]
  }
  
  @objc
  func increment(){
    count += 1
    print("count is  \(count)")
    sendEvent(withName: "onIncrement", body: ["count": count])
  }
  
  
  override func supportedEvents() -> [String]! {
    return ["onIncrement"]
  }
  
  
  // https://www.codingem.com/swift-underscore/
  // Details on how _ can be used in Swift.
  
  @objc
  func getCount(_ callback: RCTResponseSenderBlock){
    callback([count])
  }
  
  @objc
  func decrement(
    _ resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) -> Void {
    if (count == 0){
      let error = NSError(domain: " ", code: 200, userInfo: nil)
      reject("E_Count","Count cannot be negative", error)
    } else {
      count -= 1
      resolve("count was decremented : \(count)")
    }
  }
}


